src/services/completions.ts

```diff
@@ -1855,6 +1863,13 @@ function createCompletionEntry(
         }
     }
 
+    // TODO: the output code is still invalid (QualifiedName current cannot have form in namespace["invalid text name"] and but is no other way to access it)
+    // generating namespace["invalid text name"] with a type error is better than generating 'namespace.invalid text name' and have a syntax error.
+    if (needsConvertPropertyAccess && contextToken?.kind === SyntaxKind.DotToken && contextToken.parent.kind === SyntaxKind.QualifiedName) {
+        replacementSpan = createTextSpanFromNode(contextToken, sourceFile);
+        insertText = `[${quotePropertyName(sourceFile, preferences, name)}]`;
+    }

```

Comment:
Is it worth doing this if it's guaranteed to be wrong, compared to just not offering this?

Reply:
Yes, and I hope QualifiedName can contain `["stringLiteral"]` in the future, since now a type name may not be an identifier.

Reply:
What does the error message say when you land in this position? IMO it’s very annoying and confusing for completions to be invalid. It might be ok in this edge case if the diagnostic explains what’s going on and how to fix it, but if it’s just `Module 'foo' has no export "non identifier"`, then I think we need to do better somehow, and I’d lean toward not offering the completion in the meantime.

Reply:
Code:

```ts
type z = x.
```

Before this PR:

```ts
type z = x.a b
// Namespace '...' has no exported member 'a'.
// ';' expected.
// Cannot find name 'b'.
```

After this PR:

```ts
type z = x['a b']
// Cannot use namespace 'x' as a type.
```

I think the error message becomes better

Reply:
But is that better than not offering the broken completion at all? I’m not sure.